File: 2024/Day 20/Python/main.py
"""Advent of Code - 2024 - Day 20"""
from collections import deque
from itertools import combinations

Vec = tuple[int, int]
VecQueue = deque[Vec]
DistDict = dict[Vec, int]

def bfs() -> DistDict:
    "breadth first search"
    def neighbors(n: Vec):
        for direction in [(0,1),(1,0),(0,-1),(-1,0)]:
            new = (n[0] + direction[0], n[1] + direction[1])
            if 0 <= new[0] < w and 0 <= new[1] < h and maze[new[1]][new[0]] != '#':
                yield new

    q: VecQueue = deque()
    q.append(end)
    dist: DistDict = {end: 0}

    while q:
        current = q.popleft()
        for neighbor in neighbors(current):
            if neighbor not in dist:
                dist[neighbor] = dist[current] + 1
                q.append(neighbor)
    return dist

def manhattan(a: Vec, b: Vec) -> int:
    """returns the manhattan distance between a and b"""
    return abs(a[0]-b[0]) + abs(a[1]-b[1])

def saves(cheat_len: int, dist: DistDict, a: Vec, b: Vec):
    """calculates if the cheat saves enough time"""
    if dist[a] < dist[b]:
        a, b = b, a
    if dist[a] - dist[b] >= MIN_SAVE:
        m = manhattan(a, b)
        if m <= cheat_len:
            if dist[a] - dist[b] - m >= MIN_SAVE:
                return True
    return False

with open(0, encoding="utf-8") as f:
    maze = f.read().splitlines()

h = len(maze)
w = len(maze[0])

end = (-1, -1)

for y, line in enumerate(maze):
    for x, c in enumerate(line):
        if c == 'E':
            end = (x, y)
res: DistDict = bfs()
MIN_SAVE = 100
print(sum(1 for (a, b) in combinations(res, 2) if saves(2, res, a, b)))
print(sum(1 for (a, b) in combinations(res, 2) if saves(20, res, a, b)))

File: 2024/Day 20/C++/main.cpp
#include <iostream>
#include <string>
#include <vector>
#include <unordered_map>
#include <queue>
#include <limits>

template <class A, class B>
struct std::hash<std::pair<A, B>>
{
    size_t operator()(const pair<A, B> &p) const
    {
        return ((hash<A>{}(p.first) << 1) | (hash<A>{}(p.first) >> (std::numeric_limits<size_t>::digits - 1))) ^ hash<B>{}(p.second);
    }
};

template <typename T, typename U>
std::pair<T, U> operator+(const std::pair<T, U> &l, const std::pair<T, U> &r)
{
    return {l.first + r.first, l.second + r.second};
}

using Vec = std::pair<int, int>;

const Vec dirs[] = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
std::vector<Vec> get_neighbors(std::vector<std::string> &maze, Vec n, int w, int h)
{
    std::vector<Vec> neighbors;
    for (Vec direction : dirs)
    {
        Vec next = n + direction;
        if (0 <= next.first && next.first < w && 0 <= next.second && next.second < h && maze[next.second][next.first] != '#')
        {
            neighbors.push_back(next);
        }
    }
    return neighbors;
}

std::unordered_map<Vec, int> bfs(Vec end, std::vector<std::string> &maze, int w, int h)
{
    std::queue<Vec> q;
    q.push(end);
    std::unordered_map<Vec, int> dist = {{end, 0}};
    while (!q.empty())
    {
        Vec current = q.front();
        q.pop();
        for (Vec neighbor : get_neighbors(maze, current, w, h))
        {
            if (dist.find(neighbor) == dist.end())
            {
                dist[neighbor] = dist[current] + 1;
                q.push(neighbor);
            }
        }
    }
    return dist;
}

int manhattan(Vec a, Vec b)
{
    return abs(a.first - b.first) + abs(a.second - b.second);
}

const int MIN_SAVE = 100;
bool saves(int cheat_len, std::unordered_map<Vec, int> &dist, Vec a, Vec b)
{
    if (dist[a] < dist[b])
    {
        Vec tmp = a;
        a = b;
        b = tmp;
    }
    if (dist[a] - dist[b] >= MIN_SAVE)
    {
        int m = manhattan(a, b);
        if (m <= cheat_len)
        {
            if (dist[a] - dist[b] - m >= MIN_SAVE)
            {
                return true;
            }
        }
    }
    return false;
}

int solve(int cheat_len, std::unordered_map<Vec, int> &dist)
{
    std::vector<Vec> positions;
    for (auto kp : dist)
    {
        positions.push_back(kp.first);
    }
    int out = 0;
    for (int i = 0; i < positions.size()-1; i++)
    {
        for (int j = i + 1; j < positions.size(); j++)
        {
            if (saves(cheat_len, dist, positions[i], positions[j]))
            {
                out++;
            }
        }
    }
    return out;
}

int main(int argc, char *argv[])
{
    std::string line;
    std::vector<std::string> maze;
    Vec end;
    int y = 0;
    while (std::getline(std::cin, line))
    {
        for (int x = 0; x < line.size(); x++)
        {
            if (line[x] == 'E')
            {
                end = {x, y};
            }
        }
        maze.push_back(line);
        y++;
    }
    int h = maze.size();
    int w = maze[0].size();
    std::unordered_map<Vec, int> dist = bfs(end, maze, w, h);
    std::cout << solve(2, dist) << "\n";
    std::cout << solve(20, dist) << "\n";
}